### ISSUE TITLE
directory permissions incorrectly set before the subtree has been fully copied

### ISSUE THREAD

**Issue by akesandgren** [hpc/mpifileutils#400]
Using version 0.10.1

When I have a directory in the source tree which isn't writable, the target directory is created without write permission.
This is wrong.
The target dir must be created with write permission so that the files and subdirs can be copied/created.
Directory permissions must not be set/preserved until the subtree has finished copying

```
constant:
total 56
dr-xr-x---+ 3 ake folk 4096 Apr  3  2017 boundaryData/
-r--r-x---+ 1 ake folk 1546 Jul 20  2017 chemistryProperties*
-r--r-x---+ 1 ake folk 1109 Jul 20  2017 combustionProperties*

constant/boundaryData:
total 316
dr-xr-x---+ 5490 ake folk 319488 Jul 20  2017 inlet1/
constant/boundaryData/inlet1:
total 22096
dr-xr-x---+ 2 ake folk  4096 Jul 20  2017 0.0/
dr-xr-x---+ 2 ake folk  4096 Apr  3  2017 0.0001038898/
dr-xr-x---+ 2 ake folk  4096 Apr  3  2017 0.0001082185/
```

A normal cp -r or rsync -a has no problem with this, it only sets directory permissions AFTER finishing the subtree.
dcp --preserve and dsync both fail with permission denied when doing mkdir in the target tree.

Note that this only happens when running larger jobs on larger trees.

Small trees with just a few dirs have no problem.

**Comment by akesandgren** [comment]
Never mind, looks like ACL problems against our new Lustre server...

**Comment by akesandgren** [comment]
Problem seem to be in mfu_copy_xattrs when called from mfu_create_directory.
The xattrs contain system.posix_acl_default and system.posix_acl_access.
When chmod -w has been done on the src dir, the system.posix_acl_access xattr contains user::r-x resulting in write-protected target.

I suggest filtering out any system.posix_acl xattrs in mfu_copy_xattrs, like in #401

**Comment by daltonbohning** [comment]
I think part of the issue here is that `mfu_copy_xattrs` is called immediately after creating each directory/file, to accommodate Lustre striping params:
https://github.com/daltonbohning/mpifileutils/blob/4ec784108d066abe5060ebb197c1dba83e88bd7d/src/common/mfu_flist_copy.c#L933-L942

In contrast, ownership, permissions, and timestamps are copied _after_ copying the files, and starting from the deepest level, to handle similar issues with standard permissions.

I wonder if there is a solution that will allow the `system.posix_acl*` xattrs to still be copied, while also circumventing this permission issue during directory creation.

**Comment by akesandgren** [comment]
Well if they can be edited so that you have u+w then it would be ok to copy them at that point in time.

**Comment by daltonbohning** [comment]
_**Ideally**_, a full solution might look something like:
```
// copy lustre xattrs before copying files
setxattr(dst_dir) for each "lustre.*" xattr

// copy the files

// copy all other xattrs
setxattr(dst_dir) for all other non-lustre xattr
```

Or, we could **_filter out_** `system.*` attrs in the first pass, and then the second pass would add those in.

But this would require two passes over the xattrs and some filtering, which would affect performance. But perhaps this performance hit could be mitigated somewhat by using flags in some way to determine if one or both passes are necessary.
@adammoody Do you have any thoughts on this? DAOS also utilizes the `system.posix_acl*` attributes, so just not copying those at all would not be desirable.

**Comment by adilger** [comment]
@daltonbohning typically, system xattrs should not be copied directly, since they represent internal filesystem state. ACLs and such should be copied via the appropriate APIs, see GNU tar, for example.

**Comment by daltonbohning** [comment]
@adilger On my system at least, `cp --preserve=xattr src dst` does seem to copy `xattrs` set with `setfacl`
(CentOS Linux release 7.8.2003)

Running `strace cp --preserve src dst`:
```
...
fgetxattr(3, "system.posix_acl_access", ...)
fsetxattr(4, "system.posix_acl_access", ...)
...
```
So `cp` does explicitly copy over the ACLs through `xattr` access.

I think the underlying issue with `dcp` isn't that the `system.* xattrs` are being copied incorrectly (I.e. with incorrect API), but that they are just being copied at the wrong time. I think it could be possible to implement a sort of "two-pass" `xattr` copy, which would be compatible with both Lustre (needs `lustre.* xattrs` **_before_** copying files) and other setups such as yours (need `system.* xattrs` _**after**_ copying files)

Also, something that isn't yet supported is `--preserve=[ATTR_LIST]`, which would allow specifically copying only some things (mode, ownership, timestamps, xattrs)

**Comment by adammoody** [comment]
Yes, I get the impression that full xattr support needs to be generalized in a couple ways.  It seems like some xattrs must be copied before and some settings should be copied after.  We've also talked about ways to allow a user to filter out some xattr settings completely.

Related issues:
https://github.com/hpc/mpifileutils/issues/324
https://github.com/hpc/mpifileutils/issues/49

**Comment by daltonbohning** [comment]
Ah, yes! That would be a more complete solution.

**Comment by adilger** [comment]
I think there are a couple of options in this case:
- add `--xattr-include=<regexp>` and `--xattr-exclude=<regexp>` options (allow multiple) to filter xattrs
- use `/etc/xattr.conf` to filter xattrs, which is what `libattr.so` checks

I can't find any documentation for `/etc/xattr.conf` in a man page, but the [Git repo](https://git.savannah.nongnu.org/git/attr.git) has an example file like:
    # /etc/xattr.conf
    #
    # Format:
    # <pattern> <action>
    #
    # Actions:
    #   permissions - copy when trying to preserve permissions.
    #   skip - do not copy.
    system.nfs4_acl                 permissions
    system.nfs4acl                  permissions
    system.posix_acl_access         permissions
    system.posix_acl_default        permissions
    trusted.SGI_ACL_DEFAULT         skip            # xfs specific
    trusted.SGI_ACL_FILE            skip            # xfs specific
    trusted.SGI_CAP_FILE            skip            # xfs specific
    trusted.SGI_DMI_*               skip            # xfs specific
    trusted.SGI_MAC_FILE            skip            # xfs specific
    xfsroot.*                       skip            # xfs specific; obsolete
    user.Beagle.*                   skip            # ignore Beagle index data
    security.evm                    skip            # may only be written by kernel
